### ISSUE TITLE
How to select different batch samples in each iteration?

### ISSUE THREAD

**Issue by LinZichuan** [torchnet/torchnet#60]
The training code is like:
```
engine:train{
	network = model,
	iterator = getIterator('train'),
	criterion = criterion,
	optimMethod = optim.sgd,
	config = tablex.deepcopy(cnnopt),
	maxepoch = cnnopt.max_epoch,
}
```
The getIterator() function is following and only called once:
```
local function getIterator(mode)
	return tnt.ParallelDatasetIterator{
		nthread = 1,
		init = function()
			require 'torchnet'
			require 'image'
			require 'nn'
		end,
		closure = function()
			local dataset = provider[mode..'Data']

			local list_dataset = tnt.ListDataset{
				list = torch.range(1, dataset.labels:numel()):long(),
				load = function(idx)
					return {
						input = dataset.data[idx]:float(),
						target = torch.LongTensor{dataset.labels[idx]},
					}
				end,
			}
			if mode == 'train' then
				return list_dataset
				:transform{
					input = tnt.transform.compose{
						cnnopt.hflip and hflip,
						cnnopt.randomcrop > 0 and randomcrop,
					}
				}
				:batch(cnnopt.batchSize, 'skip-last')
			elseif mode == 'test' then
				return list_dataset
				:batch(cnnopt.batchSize, 'include-last')
			end
		end
	}
end
```
It seems that the DatasetIterator() will return a fixed list of batches for iterating. If I want to select batch samples at each iteration, how should I change the code? or are there other interfaces I should use?

**Comment by LinZichuan** [comment]
The training code is like:
```
engine:train{
	network = model,
	iterator = getIterator('train'),
	criterion = criterion,
	optimMethod = optim.sgd,
	config = tablex.deepcopy(cnnopt),
	maxepoch = cnnopt.max_epoch,
}
```
The getIterator() function is following and only called once:
```
local function getIterator(mode)
	return tnt.ParallelDatasetIterator{
		nthread = 1,
		init = function()
			require 'torchnet'
			require 'image'
			require 'nn'
		end,
		closure = function()
			local dataset = provider[mode..'Data']

			local list_dataset = tnt.ListDataset{
				list = torch.range(1, dataset.labels:numel()):long(),
				load = function(idx)
					return {
						input = dataset.data[idx]:float(),
						target = torch.LongTensor{dataset.labels[idx]},
					}
				end,
			}
			if mode == 'train' then
				return list_dataset
				:transform{
					input = tnt.transform.compose{
						cnnopt.hflip and hflip,
						cnnopt.randomcrop > 0 and randomcrop,
					}
				}
				:batch(cnnopt.batchSize, 'skip-last')
			elseif mode == 'test' then
				return list_dataset
				:batch(cnnopt.batchSize, 'include-last')
			end
		end
	}
end
```
It select batch samples in each iteration by a fix policy. If I want to change this policy at each iteration, how should I change the code? or are there other interfaces I should use?

**Comment by lvdmaaten** [comment]
You can wrap your `list_dataset` in a `tnt.ShuffleDataset`, and apply the batching on that shuffled dataset. Now you can can `resample()` on the `tnt.ShuffleDataset` at the end of every training epoch (for instance, in the `engine.hooks.onEndEpoch` hook.
